rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp

```diff
@@ -0,0 +1,198 @@
+    }
+
+    auto & msg = next_messages[i];
+    if (msg == nullptr) {
+      continue;
+    }
+    if (earliest_msg == nullptr || msg->time_stamp < earliest_msg->time_stamp) {
+      earliest_msg = msg;
+      earliest_msg_index = i;
+    }
+  }
+
+  // clear returned message from queue before returning it, so it can be refilled next time
+  if (earliest_msg != nullptr) {
+    next_messages[earliest_msg_index].reset();
+  }
+  return earliest_msg;
+}
+
+/// Discover what topics are in the inputs, filter out topics that can't be processed,
+/// create_topic on Writers that will receive topics.
+/// Return a map f topic -> vector of which Writers want to receive that topic,
+/// based on the RecordOptions.
+/// The output vector has bare pointers to the uniquely owned Writers,
+/// so this may not outlive the output_bags Writers.
+std::unordered_map<std::string, std::vector<rosbag2_cpp::Writer *>>
+setup_topic_filtering(
+  const std::vector<std::unique_ptr<rosbag2_cpp::Reader>> & input_bags,
+  const std::vector<
+    std::pair<std::unique_ptr<rosbag2_cpp::Writer>, rosbag2_transport::RecordOptions>
+  > & output_bags)
+{
+  std::unordered_map<std::string, std::vector<rosbag2_cpp::Writer *>> filtered_outputs;
+  std::map<std::string, std::vector<std::string>> input_topics;
+  std::unordered_map<std::string, YAML::Node> input_topics_qos_profiles;
+  std::unordered_map<std::string, std::string> input_topics_serialization_format;
+
+  for (const auto & input_bag : input_bags) {
+    auto bag_topics_and_types = input_bag->get_all_topics_and_types();
+    for (const auto & topic_metadata : bag_topics_and_types) {
+      const std::string & topic_name = topic_metadata.name;
+      input_topics.try_emplace(topic_name);
+      input_topics[topic_name].push_back(topic_metadata.type);
+      input_topics_serialization_format[topic_name] = topic_metadata.serialization_format;
+
+      // Gather all offered qos profiles from all inputs
+      input_topics_qos_profiles.try_emplace(topic_name);
+      YAML::Node & all_offered = input_topics_qos_profiles[topic_name];
+      YAML::Node offered_qos_profiles = YAML::Load(topic_metadata.offered_qos_profiles);
+      for (auto qos : offered_qos_profiles) {
+        all_offered.push_back(qos);
+      }
+    }
+  }
+
+  for (const auto & [writer, record_options] : output_bags) {
+    rosbag2_transport::TopicFilter topic_filter{record_options};
+    auto filtered_topics_and_types = topic_filter.filter_topics(input_topics);
+
+    // Done filtering - set up writer
+    for (const auto & [topic_name, topic_type] : filtered_topics_and_types) {
+      rosbag2_storage::TopicMetadata topic_metadata;
+      topic_metadata.name = topic_name;
+      topic_metadata.type = topic_type;
+
+      // Take source serialization format for the topic if output format is unspecified
+      if (record_options.rmw_serialization_format.empty()) {
+        topic_metadata.serialization_format = input_topics_serialization_format[topic_name];
```

Comment:
In the rare case that the same topic has different serialization formats in the inputs, this is the "last" serialization format that occurs in the previous loop right? Maybe note somewhere that this is the expected selection behaviour?

Reply:
Added a note comment in the docstring in the header.